File: packages/swift/Sources/ReallyMeCrypto/P256Ecdh.swift
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

import CryptoKit
import Foundation

/// P-256 ECDH backed by CryptoKit.
///
/// Public keys at the SDK boundary are compressed SEC1. The primitive returns
/// the raw 32-byte ECDH x-coordinate; protocols must apply their own labelled
/// KDF rather than relying on this low-level helper to choose one.
public enum ReallyMeP256Ecdh {
    public static let secretKeyLength = 32
    public static let compressedPublicKeyLength = 33
    public static let sharedSecretLength = 32

    public static func generateKeyPair() throws -> (publicKey: [UInt8], secretKey: [UInt8]) {
        let privateKey = P256.KeyAgreement.PrivateKey()
        return (
            publicKey: Array(privateKey.publicKey.compressedRepresentation),
            secretKey: Array(privateKey.rawRepresentation)
        )
    }

    public static func deriveKeyPair(secretKey: [UInt8]) throws -> (publicKey: [UInt8], secretKey: [UInt8]) {
        (
            publicKey: try derivePublicKey(secretKey: secretKey),
            secretKey: secretKey
        )
    }

    public static func derivePublicKey(secretKey: [UInt8]) throws -> [UInt8] {
        guard secretKey.count == secretKeyLength else {
            throw ReallyMeCryptoError.invalidInput
        }
        do {
            let privateKey = try P256.KeyAgreement.PrivateKey(rawRepresentation: Data(secretKey))
            return Array(privateKey.publicKey.compressedRepresentation)
        } catch {
            throw ReallyMeCryptoError.invalidInput
        }
    }

    public static func deriveSharedSecret(publicKey: [UInt8], secretKey: [UInt8]) throws -> [UInt8] {
        guard publicKey.count == compressedPublicKeyLength,
              secretKey.count == secretKeyLength
        else {
            throw ReallyMeCryptoError.invalidInput
        }
        do {
            let privateKey = try P256.KeyAgreement.PrivateKey(rawRepresentation: Data(secretKey))
            let peerPublicKey = try P256.KeyAgreement.PublicKey(compressedRepresentation: Data(publicKey))
            let sharedSecret = try privateKey.sharedSecretFromKeyAgreement(with: peerPublicKey)
            let bytes = sharedSecret.withUnsafeBytes { buffer in
                Array(buffer)
            }
            guard bytes.count == sharedSecretLength else {
                throw ReallyMeCryptoError.providerFailure
            }
            return bytes
        } catch let error as ReallyMeCryptoError {
            throw error
        } catch {
            throw ReallyMeCryptoError.invalidInput
        }
    }
}

File: packages/ts/src/p256Ecdh.ts
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

import { p256 } from "@noble/curves/nist.js";
import { ReallyMeCryptoError } from "./errors.js";
import { bestEffortClear } from "./memory.js";

/**
 * P-256 ECDH backed by @noble/curves.
 *
 * Public keys at the SDK boundary are compressed SEC1. The primitive returns
 * the raw 32-byte ECDH x-coordinate; protocols must apply their own labelled
 * KDF rather than relying on this low-level helper to choose one.
 */
export const P256_ECDH_SECRET_KEY_LENGTH = 32;
export const P256_ECDH_COMPRESSED_PUBLIC_KEY_LENGTH = 33;
export const P256_ECDH_SHARED_SECRET_LENGTH = 32;

export const ReallyMeP256Ecdh = {
  generateKeyPair(): { publicKey: Uint8Array; secretKey: Uint8Array } {
    const secretKey = p256.utils.randomSecretKey();
    return {
      publicKey: p256.getPublicKey(secretKey, true),
      secretKey,
    };
  },

  deriveKeyPair(secretKey: Uint8Array): { publicKey: Uint8Array; secretKey: Uint8Array } {
    return {
      publicKey: this.derivePublicKey(secretKey),
      secretKey: secretKey.slice(),
    };
  },

  derivePublicKey(secretKey: Uint8Array): Uint8Array {
    if (secretKey.length !== P256_ECDH_SECRET_KEY_LENGTH) {
      throw new ReallyMeCryptoError("invalid-input");
    }
    try {
      return p256.getPublicKey(secretKey, true);
    } catch {
      throw new ReallyMeCryptoError("invalid-input");
    }
  },

  deriveSharedSecret(publicKey: Uint8Array, secretKey: Uint8Array): Uint8Array {
    if (
      publicKey.length !== P256_ECDH_COMPRESSED_PUBLIC_KEY_LENGTH ||
      secretKey.length !== P256_ECDH_SECRET_KEY_LENGTH
    ) {
      throw new ReallyMeCryptoError("invalid-input");
    }
    try {
      const uncompressed = p256.getSharedSecret(secretKey, publicKey, false);
      const sharedSecret = uncompressed.slice(1, 33);
      bestEffortClear(uncompressed);
      return sharedSecret;
    } catch {
      throw new ReallyMeCryptoError("invalid-input");
    }
  },
} as const;
